### ISSUE TITLE
Performance is wrose with jemalloc 5.3

### ISSUE THREAD

**Issue by CaoE** [jemalloc/jemalloc#2395]
When update jemalloc 5.2 to jemalloc 5.3, the perforamce of PyTorch is worse.
Take efficientnet_b6 as a example: 
- Using jemalloc 5.3.0
<img width="719" alt="image" src="https://user-images.githubusercontent.com/23565213/223930782-c90db66c-05c0-4374-8fa1-1c875a683ed8.png">

- Using jemalloc 5.2.1
<img width="704" alt="image" src="https://user-images.githubusercontent.com/23565213/223930876-bb6b5025-6c26-4f4a-a104-d48b4b38ac89.png">

**Comment by interwq** [comment]
Thanks for sharing the findings. This is certainly unexpected. Am I understanding correctly that the total CPU time more than tripled? If you could share the malloc stats, it should be able to point out what went wrong. You can make it output by the end using e.g. "MALLOC_CONF=stats_print:true ./binary" we would want the stats from both sides to compare.

**Comment by CaoE** [comment]
> Am I understanding correctly that the total CPU time more than tripled?

The results are obtained by torch profiler, it may introduce overhead, so the gap is not so big without torch profiler. The throughput gap is still about 60% without profiler. 

> You can make it output by the end using e.g. "MALLOC_CONF=stats_print:true ./binary" we would want the stats from both sides to compare.

- The malloc stats with jemalloc 5.2.1 on a socket with 28 cores:
[jemalloc-5.2.log](https://github.com/jemalloc/jemalloc/files/10940330/jemalloc-5.2.log)

- The malloc stats with jemalloc 5.3.0 on a socket with 28 cores (same machine):
[jemalloc-5.3.log](https://github.com/jemalloc/jemalloc/files/10940329/jemalloc-5.3.log)

**Comment by interwq** [comment]
@CaoE I notice that the output has:
from 5.2.1
```
-------------------- Summary --------------------
Latency:	 303.905 ms
Throughput:	 3.29 samples/s
Latency P50:	 301.592 ms
Latency P90:	 313.361 ms
Latency P99:	 327.857 ms
```

VS
5.3
```
 -------------------- Summary --------------------
Latency:	 303.315 ms
Throughput:	 3.30 samples/s
Latency P50:	 303.749 ms
Latency P90:	 311.193 ms
Latency P99:	 314.768 ms
```

Does this mean this test result happen to be neutral? Asking because the malloc_stats look very similar -- so close that I'd be a bit surprised if the overall efficiency was significantly different.

The only notable difference from the malloc stats, is that the thread cache refill happens a bit more frequently with 5.3 (we changed the way thread cache GC is scheduled). However that should be a rather minor factor when looking at the overall performance, especially since there's no lock contention observed.

**Comment by CaoE** [comment]
Hi @interwq When `stats_print:true` is set, the performance i.e. throughput decreased. Is this expected ?

**Comment by interwq** [comment]
@CaoE Not really expected. That option only triggers the output at exit. When the program is running, the stats are maintained regardless of the option. Unless the binaries are launched many many times, which triggered the stats many times as well?

Do you mean that the throughput decrease for 5.2.1, or both?

**Comment by CaoE** [comment]
> Unless the binaries are launched many many times, which triggered the stats many times as well?

Oh, yes, the binaries are launched many times including warmup. So should the binary only be launched once ?

**Comment by interwq** [comment]
So printing to the terminal certainly takes time. If the run itself takes a short time, then the cost from the prints could take a non trivial percentage. Maybe try adding stats_print_opts:a which should omit the per arena stats -- this should shorten the stats significantly. 

But yeah if you could, only apply the stats option to one binary run after warm up would be good.

**Comment by CaoE** [comment]
@interwq I re-measured the performance with stats_print:true, which show s a performance gap  :
- The malloc stats with jemalloc 5.2.1 on a socket with 28 cores:
[jemallc_5_2.log](https://github.com/jemalloc/jemalloc/files/10976848/jemallc_5_2.log)

- The malloc stats with jemalloc 5.3.0 on a socket with 28 cores (same machine):
[jemalloc_5_3.log](https://github.com/jemalloc/jemalloc/files/10976849/jemalloc_5_3.log)

**Comment by interwq** [comment]
@CaoE the stats do show a big difference in the page cache: 5.3 purged a lot more pages and has much lower memory usage, while 5.2 keeps almost all dirty pages cached. 

From the config, it looks you prioritize CPU a lot more than memory? If so, can you try setting  dirty_decay_ms:-1 ? It could help to confirm the difference is from purging. 

One potential cause could be from the oversize_threshold handling. I'm suspecting the 0 value isn't handled properly. Can you please also try oversize_threshold:1000000000 with both versions (but without the decay change above) ? If the difference is gone then it can be confirmed. 

Thanks again for help testing!

**Comment by CaoE** [comment]
Thanks for your help.  When set dirty_decay_ms:-1 the gap is gone: [test.log](https://github.com/jemalloc/jemalloc/files/10977247/test.log).
We set `ALLOC_CONF="oversize_threshold:1,background_thread:true,metadata_thp:auto,dirty_decay_ms:-1,muzzy_decay_ms:9000000000"`.  
Is this the best config ? Could you please share us a BKC for  jemalloc 5.3 ?

**Comment by interwq** [comment]
We made another tweak to the `oversize_threshold` feature recently: #2466

You can try the current dev branch, without the `oversize_threshold:NNN` setting in MALLOC_CONF. This should be the best balance between CPU and memory (as long as you include `background_thread:true`).

**Comment by dgriff67** [comment]
Do you have timeline for when will this fix might be released?

Further when it might be adopted on to a Debian bookworm image. We have experienced a 10% - 15% deterioration in our training / prediction with the change of version 5.2.1-3 on bullseye to 5.3.0-1 on bookworm.

Some idea on timeline would be great as we are looking to upgrade to bookworm soon for security reasons, but cannot sacrifice performance.
